File: ts/src/visual-graph/builder/build-scope.ts
import { DEFINITION_TYPE } from "../../analyzer/definition-type.js";
import type { SerializedScope } from "../../ir/serialized/serialized-scope.js";
import { NODE_KIND } from "../node-kind.js";
import { VISUAL_ELEMENT_TYPE } from "../visual-element-type.js";
import type { VisualElement } from "../visual-element.js";
import type { VisualNode } from "../visual-node.js";
import type { VisualSubgraph } from "../visual-subgraph.js";
import { buildChildren } from "./build-children.js";
import type { BuildState } from "./build-state.js";
import type { BuilderContext } from "./context.js";
import { describeSubgraph } from "./describe-subgraph.js";
import { ensureBeyondDepthStub } from "./ensure-beyond-depth-stub.js";
import { isCollapsed } from "./is-collapsed.js";
import { attachLoopTestAnchor } from "./loop-test-anchor.js";
import { makeVariableNode } from "./make-variable-node.js";
import { nodeId } from "./node-id.js";
import { shouldSubgraph } from "./should-subgraph.js";
import { attachSwitchDiscriminantAnchor } from "./switch-discriminant-anchor.js";
import { visibleAncestorSubgraph } from "./visible-ancestor-subgraph.js";
import { writeOpNodeId } from "./write-op-node-id.js";

type Container = Readonly<{
  elements: /* mutable */ VisualElement[];
}>;

function recordCollapsedDescendants(
  scope: SerializedScope,
  rootScopeId: string,
  ctx: BuilderContext,
  state: BuildState,
): void {
  state.collapsedRootByScope?.set(scope.id, rootScopeId);
  for (const childId of scope.childScopes) {
    const child = ctx.scopeMap.get(childId);
    if (!child) {
      continue;
    }
    recordCollapsedDescendants(child, rootScopeId, ctx, state);
  }
}

export function buildScope(
  scope: SerializedScope,
  container: Container,
  ctx: BuilderContext,
  state: BuildState,
): void {
  if (isCollapsed(scope, ctx.depths)) {
    // Nothing rendered for the collapsed subtree itself. Anchor for
    // cross-boundary edges, in priority order:
    //   1. the owning variable of the collapsed scope (e.g. `fnB` for
    //      `function fnB() { ... }`) -- already emitted in the parent;
    //   2. otherwise a single BeyondDepth `((...))` stub placed inside
    //      the closest visible ancestor subgraph, so anonymous callbacks
    //      / branch / loop / try bodies / bare blocks announce "the
    //      read happens somewhere inside the surviving outer container"
    //      via a circle marker that mirrors the pruning boundary stub.
    recordCollapsedDescendants(scope, scope.id, ctx, state);
    const ownerVarId = ctx.subgraphOwnerVar.get(scope.id) ?? null;
    let anchorId: string | null;
    if (ownerVarId !== null) {
      anchorId = nodeId(ownerVarId);
    } else {
      const parentSg = visibleAncestorSubgraph(scope, ctx, state);
      anchorId = parentSg ? ensureBeyondDepthStub(parentSg, state) : null;
    }
    if (anchorId !== null) {
      state.collapsedAnchorByRoot?.set(scope.id, anchorId);
      // If the collapsed scope is the body of a control statement, its
      // test anchor (if-test, for-test, while-test, switch discriminant)
      // could not be created -- route reads of that predicate to the
      // same stub / variable anchor so they share the boundary marker.
      const blockCtx = scope.blockContext;
      if (blockCtx && state.suppressedPredicateRedirect) {
        state.suppressedPredicateRedirect.set(
          blockCtx.parentSpanOffset,
          anchorId,
        );
      }
    }
    return;
  }

  const subgraphHere = shouldSubgraph(scope);
  let bodyContainer: Container = container;
  let bodySubgraph: VisualSubgraph | null = null;
  if (subgraphHere) {
    const sg = describeSubgraph(scope, ctx.subgraphOwnerVar, ctx.variableMap);
    container.elements.push(sg);
    bodyContainer = sg;
    bodySubgraph = sg;
    state.subgraphByScope.set(scope.id, sg);
    const ownerVar = ctx.subgraphOwnerVar.get(scope.id);
    if (ownerVar) {
      state.functionSubgraphByFn.set(ownerVar, sg);
    }
  }
  for (const vid of scope.variables) {
    const v = ctx.variableMap.get(vid);
    if (!v) {
      continue;
    }
    const node = makeVariableNode(v);
    state.nodeIdOriginScope?.set(node.id, scope.id);
    bodyContainer.elements.push(node);
  }
  const ops = ctx.writeOpsByScope.get(scope.id) ?? [];
  for (const op of ops) {
    const ownerVar = ctx.variableMap.get(op.varId);
    const ownerDef = ownerVar?.defs[0];
    const declarationKind =
      ownerDef?.type === DEFINITION_TYPE.Variable
        ? ownerDef.declarationKind
        : null;
    const node = {
      type: VISUAL_ELEMENT_TYPE.Node,
      id: writeOpNodeId(op.refId),
      kind: NODE_KIND.WriteReference,
      name: op.varName,
      line: op.line,
      endLine: null,
      isJsxElement: false,
      unused: false,
      declarationKind,
    } satisfies VisualNode;
    state.nodeIdOriginScope?.set(node.id, scope.id);
    bodyContainer.elements.push(node);
  }
  buildChildren(scope, bodyContainer, ctx, state);
  if (bodySubgraph !== null) {
    attachLoopTestAnchor(scope, bodySubgraph, state);
    attachSwitchDiscriminantAnchor(scope, bodySubgraph, state);
  }
}

File: crates/unsnarl-visual-graph/src/builder/build_scope.rs
//! Per-scope render dispatcher: if the scope is gated below the
//! depth ceiling its subtree is recorded as collapsed and the
//! visible anchor (owner variable or a BeyondDepth stub) is
//! cached so edges crossing the boundary still have a destination;
//! otherwise the scope's own subgraph (when it should have one) is
//! constructed, variables and write ops are pushed in source order,
//! children are walked, and loop-test / switch-discriminant anchors
//! are staged for end-of-build placement.

use unsnarl_ir::serialized::SerializedDefinition;
use unsnarl_ir::serialized::SerializedScope;
use unsnarl_oxc_parity::VariableDeclarationKind;

use crate::visual_element_type::NodeTypeTag;
use crate::visual_node::{SyntheticExtras, SyntheticNodeKind, SyntheticVisualNode, VisualNode};

use super::arena::{BuildArena, Container, ElementHandle, SubgraphIdx};
use super::build_children::build_children;
use super::context::BuilderContext;
use super::describe_subgraph::describe_subgraph;
use super::ensure_beyond_depth_stub::ensure_beyond_depth_stub;
use super::is_collapsed::is_collapsed;
use super::loop_test_anchor::attach_loop_test_anchor;
use super::make_variable_node::make_variable_node;
use super::node_id::node_id;
use super::should_subgraph::should_subgraph;
use super::state::BuildState;
use super::switch_discriminant_anchor::attach_switch_discriminant_anchor;
use super::visible_ancestor_subgraph::visible_ancestor_subgraph;
use super::write_op_node_id::write_op_node_id;

fn record_collapsed_descendants(
    state: &mut BuildState,
    ctx: &BuilderContext<'_>,
    scope: &SerializedScope,
    root_scope_id: &str,
) {
    state
        .collapsed_root_by_scope
        .insert(scope.id.value().to_string(), root_scope_id.to_string());
    for child_id in &scope.child_scopes {
        let Some(child) = ctx.scope_map.get(child_id.value()).copied() else {
            continue;
        };
        record_collapsed_descendants(state, ctx, child, root_scope_id);
    }
}

pub fn build_scope(
    arena: &mut BuildArena,
    state: &mut BuildState,
    ctx: &BuilderContext<'_>,
    scope: &SerializedScope,
    container: Container,
) {
    if is_collapsed(scope, ctx.depths.as_ref()) {
        record_collapsed_descendants(state, ctx, scope, scope.id.value());
        let owner_var_id = ctx.subgraph_owner_var.get(scope.id.value()).cloned();
        let anchor_id: Option<String> = if let Some(var_id) = owner_var_id.as_ref() {
            Some(node_id(var_id))
        } else {
            let parent_sg: Option<SubgraphIdx> = visible_ancestor_subgraph(scope, ctx, state);
            parent_sg.map(|sg| ensure_beyond_depth_stub(arena, state, sg))
        };
        if let Some(anchor) = anchor_id {
            state
                .collapsed_anchor_by_root
                .insert(scope.id.value().to_string(), anchor.clone());
            if let Some(block_ctx) = scope.block_context.as_ref() {
                state
                    .suppressed_predicate_redirect
                    .insert(block_ctx.parent_span_offset().0, anchor);
            }
        }
        return;
    }

    let subgraph_here = should_subgraph(scope);
    let mut body_container = container;
    let mut body_subgraph: Option<SubgraphIdx> = None;
    if subgraph_here {
        let sg_descriptor = describe_subgraph(scope, &ctx.subgraph_owner_var, &ctx.variable_map);
        let idx = arena.push_subgraph(sg_descriptor);
        arena.append_child(container, ElementHandle::Subgraph(idx));
        body_container = Container::Subgraph(idx);
        body_subgraph = Some(idx);
        state
            .subgraph_by_scope
            .insert(scope.id.value().to_string(), idx);
        if let Some(owner_var) = ctx.subgraph_owner_var.get(scope.id.value()) {
            state.function_subgraph_by_fn.insert(owner_var.clone(), idx);
        }
    }
    for vid in &scope.variables {
        let Some(v) = ctx.variable_map.get(vid.value()).copied() else {
            continue;
        };
        let node = make_variable_node(v);
        let node_id_str = node.id().to_string();
        let idx = arena.push_node(node);
        state
            .node_id_origin_scope
            .insert(node_id_str, scope.id.value().to_string());
        arena.append_child(body_container, ElementHandle::Node(idx));
    }
    let empty_ops: Vec<super::write_op::WriteOp> = Vec::new();
    let ops_slice: &[super::write_op::WriteOp] = ctx
        .write_ops_by_scope
        .get(scope.id.value())
        .unwrap_or(&empty_ops);
    let ops_owned: Vec<super::write_op::WriteOp> = ops_slice.to_vec();
    for op in &ops_owned {
        let owner_var = ctx.variable_map.get(op.var_id.as_str()).copied();
        let declaration_kind: Option<VariableDeclarationKind> = owner_var
            .and_then(|v| v.defs.first())
            .and_then(|def| match def {
                SerializedDefinition::Variable(d) => Some(d.declaration_kind().clone()),
                _ => None,
            });
        let id = write_op_node_id(&op.ref_id);
        let node = VisualNode::Synthetic(SyntheticVisualNode {
            r#type: NodeTypeTag::Node,
            id: id.clone(),
            kind: SyntheticNodeKind::WriteReference,
            name: op.var_name.clone(),
            line: op.line,
            end_line: None,
            is_jsx_element: false,
            unused: false,
            extras: SyntheticExtras::WriteOp { declaration_kind },
        });
        let node_idx = arena.push_node(node);
        state
            .node_id_origin_scope
            .insert(id, scope.id.value().to_string());
        arena.append_child(body_container, ElementHandle::Node(node_idx));
    }
    build_children(arena, state, ctx, scope, body_container);
    if let Some(sg) = body_subgraph {
        attach_loop_test_anchor(arena, state, scope, sg);
        attach_switch_discriminant_anchor(arena, state, scope, sg);
    }
}

#[cfg(test)]
#[path = "build_scope_test.rs"]
mod build_scope_test;
